src/Rules.Framework/ConditionTypeExtractor.cs

```diff
@@ -0,0 +1,76 @@
+namespace Rules.Framework
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using Rules.Framework.Core;
+    using Rules.Framework.Core.ConditionNodes;
+
+    /// <summary>
+    /// Extracts Conditions Types from a Group of Rules.
+    /// </summary>
+    /// <typeparam name="TContentType">The content type that allows to categorize rules.</typeparam>
+    /// <typeparam name="TConditionType">
+    /// The condition type that allows to filter rules based on a set of conditions.
+    /// </typeparam>
+    public class ConditionTypeExtractor<TContentType, TConditionType> : IConditionTypeExtractor<TContentType, TConditionType>
+    {
+        /// <summary>
+        /// Get the unique condition types associated with rules of a specific content type/>.
+        /// </summary>
+        /// <param name="matchedRules"></param>
+        /// <remarks>
+        /// <para>
+        /// A set of rules is requested to rules data source and all conditions are evaluated
+        /// against them to provide a set of matches.
+        /// </para>
+        /// <para>All rules matching supplied conditions are returned.</para>
+        /// </remarks>
+        /// <returns>the matched rule; otherwise, null.</returns>
+        public IEnumerable<TConditionType> GetConditionTypes(IEnumerable<Rule<TContentType, TConditionType>> matchedRules)
+        {
+            var conditionTypes = new List<TConditionType>();
+
+            if (!matchedRules.Any())
+            {
+                return conditionTypes;
+            }
+
+            foreach (var rule in matchedRules)
+            {
+                var rootCondition = rule.RootCondition;
+
+                VisitConditionNode(rootCondition, conditionTypes);
+            }
+
+            return conditionTypes;
+        }
+
+        private static void VisitConditionNode(IConditionNode<TConditionType> conditionNode, List<TConditionType> conditionTypes)
```

Comment:
I was thinking about the performance implications of using a `List<TConditionType>` and I came up with the idea, what if we a `HashSet<TConditionType>`? It has the same performance benefits as a `Dictionary`.

https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=net-6.0

What do you think about this?

Reply:
I agree, with this we don't need to check if the value already exists in the list in order to add it. I've updated the code with the change.